app/src/main/cpp/skyline/gpu/fence_cycle.h

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2021 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#pragma once
+
+#include <forward_list>
+#include <vulkan/vulkan_raii.hpp>
+#include <common.h>
+
+namespace skyline::gpu {
+    struct FenceCycle;
+
+    /**
+     * @brief Any object whose lifetime can be attached to a fence cycle needs to inherit this class
+     */
+    struct FenceCycleDependency {
+      private:
+        std::shared_ptr<FenceCycleDependency> next{}; //!< A shared pointer to the next dependendency to form a linked list
+        friend FenceCycle;
+    };
+
+    /**
+     * @brief A wrapper around a Vulkan Fence which only tracks a single reset -> signal cycle with the ability to attach lifetimes of objects to it
+     * @note This provides the guarantee that the fence must be signalled prior to destruction when objects are to be destroyed
+     * @note All waits to the fence **must** be done through the same instance of this, the state of the fence changing externally will lead to UB
+     */
+    struct FenceCycle {
+      private:
+        std::atomic_flag signalled;
+        const vk::raii::Device &device;
+        vk::Fence fence;
+        std::shared_ptr<FenceCycleDependency> list;
+
+        /**
+         * @brief Sequentially iterate through the shared_ptr linked list of dependencies and reset all pointers in a thread-safe atomic manner
+         * @note We cannot simply nullify the base pointer of the list as a false dependency chain is maintained between the objects when retained exteranlly
+         */
+        void DestroyDependencies() {
+            auto current{std::atomic_exchange_explicit(&list, std::shared_ptr<FenceCycleDependency>{}, std::memory_order_acquire)};
+            while (current) {
+                std::shared_ptr<FenceCycleDependency> next{};
+                next.swap(current->next);
+                current.swap(next);
+            }
+        }
+
+      public:
+        FenceCycle(const vk::raii::Device &device, vk::Fence fence) : signalled(false), device(device), fence(fence) {
+            device.resetFences(fence);
+        }
+
+        ~FenceCycle() {
+            Wait();
```

Comment:
If that happens then the GPU driver will likely throw an error that isn't timeout-related, we should depend on that rather than introduce an arbitrary timeout